### PR TITLE
ISPN-15322 Server iteration manager using wrong storage and request t…

### PR DESCRIPTION
**Pull request by jabolina**
…ype for filtering

https://issues.redhat.com/browse/ISPN-15322

* Iteration manager now verifies on the request media type.

### CODE REVIEW

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/impl/iteration/MediaTypeMultiServerRemoteIteratorTest.java [hunk 1]
```diff
@@ -29,6 +29,7 @@ protected void createCacheManagers() throws Throwable {
    private ConfigurationBuilder getCacheConfiguration() {
       ConfigurationBuilder builder = hotRodCacheConfiguration(getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC, false));
       if (key != null) builder.encoding().key().mediaType(key);
+      if (key != null) builder.encoding().value().mediaType(key);
```
Comment (wburns) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/impl/iteration/MediaTypeMultiServerRemoteIteratorTest.java:32]: Shouldn't this be the value media type being passed?

File: server/core/src/main/java/org/infinispan/server/iteration/DefaultIterationManager.java [hunk 2]
```diff
@@ -151,7 +151,10 @@ public IterationState start(AdvancedCache cache, BitSet segments, String filterC
       DataConversion valueDataConversion = cache.getValueDataConversion();
       Function<Object, Object> unmarshaller = p -> encoderRegistry.convert(p, requestValueType, APPLICATION_OBJECT);
 
-      MediaType storageMediaType = cache.getValueDataConversion().getStorageMediaType();
+      // The iterator converts from the request type to the storage type. The filter will receive the data already converted.
+      MediaType storageMediaType = valueDataConversion.getRequestMediaType() != null && valueDataConversion.getRequestMediaType() != MediaType.APPLICATION_UNKNOWN
```
Comment (wburns) [inline server/core/src/main/java/org/infinispan/server/iteration/DefaultIterationManager.java:155]: Hrmm... this filter stuff looks wrong. We seem to always be applying the value media type to do the conversion and don't pay attention to the key media type at all... am I missing something?

**Review by wburns — CHANGES_REQUESTED** [review]

File: server/core/src/main/java/org/infinispan/server/iteration/DefaultIterationManager.java (hunk 2 above)
Comment (jabolina) [reply server/core/src/main/java/org/infinispan/server/iteration/DefaultIterationManager.java:155]: Yeah. I noticed this as well. This method only handles the MediaType for values. If a conversion happens, the key is converted based on the value type. Let me address your other comment, and I'll mix key/value types.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/impl/iteration/MediaTypeMultiServerRemoteIteratorTest.java (hunk 1 above)
Comment (jabolina) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/impl/iteration/MediaTypeMultiServerRemoteIteratorTest.java:32]: Let me mix the key and value types and see if there is an issue.

Comment (jabolina) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/impl/iteration/MediaTypeMultiServerRemoteIteratorTest.java:32]: Any time the value is configured with `text/plain,` the test fails. I can take a closer look to see why.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/impl/iteration/BaseMultiServerRemoteIteratorTest.java [hunk 3]
```diff
@@ -214,17 +214,17 @@ public void testRetrieveMetadata() {
...
       }
 
       @ProtoName("HexFilterConverter")
-      static class HexFilterConverter extends AbstractKeyValueFilterConverter<Object, Integer, String> {
+      static class HexFilterConverter extends AbstractKeyValueFilterConverter<Object, Object, String> {
          @Override
-         public String filterAndConvert(Object key, Integer value, Metadata metadata) {
-            return Integer.toHexString(value);
+         public String filterAndConvert(Object key, Object value, Metadata metadata) {
+            return Integer.toHexString(Integer.parseInt(value.toString()));
```
Comment (wburns) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/impl/iteration/BaseMultiServerRemoteIteratorTest.java:227]: Have to admit this code does not sit well for me, but I guess if it is needed solely for tests with text/plain that is okay.

**Pull request merged**